numpyro/infer/elbo.py

```diff
@@ -73,6 +80,104 @@ def single_particle_elbo(rng_key):
             return - jnp.mean(vmap(single_particle_elbo)(rng_keys))
 
 
+def _get_log_prob_sum(site):
+    if site['intermediates']:
+        log_prob = site['fn'].log_prob(site['value'], site['intermediates'])
+    else:
+        log_prob = site['fn'].log_prob(site['value'])
+    log_prob = scale_and_mask(log_prob, site['scale'])
+    return jnp.sum(log_prob)
+
+
+def _check_mean_field_requirement(model_trace, guide_trace):
+    """
+    Checks that the guide and model sample sites are ordered identically.
+    This is sufficient but not necessary for correctness.
+    """
+    model_sites = [name for name, site in model_trace.items()
+                   if site["type"] == "sample" and name in guide_trace]
+    guide_sites = [name for name, site in guide_trace.items()
+                   if site["type"] == "sample" and name in model_trace]
+    assert set(model_sites) == set(guide_sites)
+    if model_sites != guide_sites:
+        warnings.warn("Failed to verify mean field restriction on the guide. "
+                      "To eliminate this warning, ensure model and guide sites "
+                      "occur in the same order.\n" +
+                      "Model sites:\n  " + "\n  ".join(model_sites) +
+                      "Guide sites:\n  " + "\n  ".join(guide_sites))
+
+
+class TraceMeanField_ELBO(ELBO):
+    """
+    A trace implementation of ELBO-based SVI. This is currently the only
+    ELBO estimator in NumPyro that uses analytic KL divergences when those
+    are available.
+
+    .. warning:: This estimator may give incorrect results if the mean-field
```

Comment:
is it worth including the full requirements?

```

    The mean field condition is a sufficient but not necessary condition for
    this estimator to be correct. The precise condition is that for every
    latent variable `z` in the guide, its parents in the model must not include
    any latent variables that are descendants of `z` in the guide. Here
    'parents in the model' and 'descendants in the guide' is with respect
    to the corresponding (statistical) dependency structure. For example, this
    condition is always satisfied if the model and guide have identical
    dependency structures.
```

Reply:
Thanks, I'll add it.